service_discovery/core/src/registry_impl.rs

```diff
@@ -0,0 +1,457 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+// SPDX-License-Identifier: MIT
+
+//! Module containing gRPC service implementation based on [`proto_servicediscovery::chariott_registry::v1`].
+//!
+//! Provides a gRPC endpoint for external services to interact with to register and discover
+//! services. Note: Identifiers in all Registry operations are case-sensitive.
+//!
+use parking_lot::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+use proto_servicediscovery::chariott_registry::v1::registry_server::Registry;
+use proto_servicediscovery::chariott_registry::v1::{
+    DiscoverByNamespaceRequest, DiscoverByNamespaceResponse, DiscoverServiceRequest,
+    DiscoverServiceResponse, InspectRequest, InspectResponse, RegisterRequest, RegisterResponse,
+    RegistrationStatus, ServiceMetadata, UnregisterRequest, UnregisterResponse,
+};
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::vec::Vec;
+use tonic::{Request, Response, Status};
+use tracing::{info, warn};
+
+/// Base structure for the registry gRPC service
+#[derive(Clone, Debug)]
+pub struct RegistryImpl {
+    pub registry_map: Arc<RwLock<HashMap<ServiceIdentifiers, ServiceMetadata>>>,
+}
+
+#[tonic::async_trait]
+impl Registry for RegistryImpl {
+    /// Registers a service by adding it to the registry.
+    ///
+    /// This function registers a service based on a [`RegisterRequest`]. Returns a
+    /// [`RegisterResponse`].
+    ///
+    /// # Arguments
+    ///
+    /// * `request` - A [`RegisterRequest`] wrapped by a [`tonic::Request`].
+    async fn register(
+        &self,
+        request: Request<RegisterRequest>,
+    ) -> Result<Response<RegisterResponse>, Status> {
+        let request_inner = request.into_inner();
+        let service_to_register =
+            request_inner.service.ok_or_else(|| Status::invalid_argument("Service is required"))?;
+        let service_identifiers = ServiceIdentifiers {
+            namespace: service_to_register.namespace.clone(),
+            name: service_to_register.name.clone(),
+            version: service_to_register.version.clone(),
+        };
+        info!("Received a register request for: {:?}", service_identifiers);
+
+        let registration_status;
+        // This block controls the lifetime of the lock.
+        {
+            let mut lock: RwLockWriteGuard<HashMap<ServiceIdentifiers, ServiceMetadata>> =
+                self.registry_map.write();
+            match lock.get(&service_identifiers) {
+                Some(_) => {
+                    lock.insert(service_identifiers.clone(), service_to_register.clone());
```

Comment:
because you make the same call to `lock.insert` twice, consider making this call after the match block instead

also, is the intent here to just completely discard the old entry?

Reply:
Changed the behavior so that we fail anyone trying to register an existing service. We will revisit this later to determine how we will handle the "update" case